Add CBOR/CWT encoding (#106)
vcstuff/draft-ietf-oauth-status-list: draft-ietf-oauth-status-list.md

  RFC6838: RFC6838
  RFC7515: RFC7515
  RFC7519: RFC7519
  RFC9052: RFC9052
  RFC8259: RFC8259
  RFC8392: RFC8392
  RFC8949: RFC8949
  RFC9110: RFC9110
  RFC9111: RFC9111
  IANA.MediaTypes:
    author:
      org: "IANA"
    title: "Media Types"
    target: "https://www.iana.org/assignments/media-types/media-types.xhtml"
  IANA.JOSE:
    author:
      org: "IANA"
    title: "JSON Object Signing and Encryption (JOSE)"
    target: "https://www.iana.org/assignments/jose/jose.xhtml"
  IANA.JWT:
    author:
      org: "IANA"
    title: "JSON Web Token Claims"
    target: "https://www.iana.org/assignments/jwt/jwt.xhtml"
  IANA.CWT:
    author:
      org: "IANA"
    title: "CBOR Web Token (CWT) Claims"
    target: "https://www.iana.org/assignments/cwt/cwt.xhtml"
  CWT.typ:
    author: 

>       fullname: "M. B. Jones"

Reviewer: ```suggestion
      name: "M. B. Jones"
```

--- abstract

> This specification defines status list data structures for representing the status of JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Web Tokens (CWTs) {{RFC8392}}, and other tokens secured by CBOR Object Signing and Encryption {{RFC9052}}.

Reviewer: ```suggestion
This specification defines status list data structures for representing the status of JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Web Tokens (CWTs) {{RFC8392}}, and other tokens secured by CBOR Object Signing and Encryption {{RFC9052}}.
This specification defines status list data structures and processing rules for representing the status of tokens secured by JSON Object Signing and Encryption {{IANA.JOSE}} or CBOR Object Signing and Encryption {{RFC9052}}, such as JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Web Tokens (CWTs) {{RFC8392}} and ISO mdoc.
```

The status list data structures themselves are also represented as JWTs or CWTs.

--- middle

# Introduction

> JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Object Signing and Encryption (COSE) {{RFC8152}} objects, and CBOR Web Tokens (CWTs) {{RFC8392}} as secure token formats, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time, which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered invalidated or suspended by its issuer.

Reviewer 2: That sounds like a good idea imho

> JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Object Signing and Encryption (COSE) {{RFC9052}} objects, and CBOR Web Tokens (CWTs) {{RFC8392}} as secure token formats, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time, which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered invalidated or suspended by its issuer.

Reviewer: ```suggestion
Token formats secured by JOSE {{IANA.JOSE}} or COSE {{RFC9052}, such as JSON Web Tokens (JWTs) {{RFC7519}}, CBOR Web Tokens (CWTs) {{RFC8392}} and ISO mdoc, have vast possible applications. Some of these applications can involve issuing a token whereby certain semantics about the token can change over time, which are important to be able to communicate to relying parties in an interoperable manner, such as whether the token is considered invalidated or suspended by its issuer.
```

The following example illustrates the JSON representation of the Status List:

~~~~~~~~~~
{::include ./examples/status_list_encoding_json}
~~~~~~~~~~

## Status List in CBOR Format {#status-list-cbor}

This section defines the `StatusList` structure for a CBOR-encoded Status List.

> The `StatusList` structure is a map (Major Type 5) and defines the following entries:

Reviewer: Are we considering to register a CBOR tag for StatusList itself?

Author: No, this is not needed.

Reviewer: Why not? The StatusList is a separate data structure that may live on its own, dependent from the StatusListToken. Doesn't it make sense in this case or should this be solved by media types alone?

Author: Tags are usually not needed. In all cases, the structure of the CBOR is pre-defined. That is why CBOR tags are not really needed in most cases.

## Status List Token in CWT Format {#status-list-token-cwt}

The Status List Token MUST be encoded as a "CBOR Web Token (CWT)" according to {{RFC8392}}.

> Applications MAY use the media type `statuslist+cwt` for a CWT in accordance with the rules outlined in this specification.

Author: This is what CWT says about the media type. As noted above typ header in CWT/COSE is pending.

The following content applies to the CWT Claims Set:

* `1` (issuer): REQUIRED. Same definition as `iss` claim in (#status-list-token-jwt).
* `2` (subject): REQUIRED. Same definition as `sub` claim in (#status-list-token-jwt).
* `6` (issued at): REQUIRED. Same definition as `iat` claim in (#status-list-token-jwt).
* `4` (expiration time): OPTIONAL. Same definition as `exp` claim in (#status-list-token-jwt).
* `TBD` (status list): REQUIRED. The status list claim MUST specify the Status List conforming to the rules outlined in [](#status-list-cbor).

The following additional rules apply:

1. The CWT MAY contain other claims.

2. The CWT MUST be digitally signed using an asymmetric cryptographic algorithm. Relying parties MUST reject the JWT if it is using a Message Authentication Code (MAC) algorithm. Relying parties MUST reject CWTs with an invalid signature.

3. Relying parties MUST reject CWTs that are not valid in all other respects per "CBOR Web Token (CWT)" {{RFC8392}}.

4. Application of additional restrictions and policy are at the discretion of the verifying party.

The following is a non-normative example for a Status List Token in CWT format:

> ~~~~~~~~~~

Author: Yep, the typ header is not defined in IANA yet. There is draft for that though. I can add the typ header but the label (14) is not confirmed yet.

> ~~~~~~~~~~

Author: Content type is about the secured payload which could be statuslist+cbor. typ could be statulist+cwt. Both can be used.

* `14` (type): REQUIRED. The type of the CWT MUST be `statuslist+cwt` as defined in {{CWT.typ}}. 

The following content applies to the CWT Claims Set:

* `1` (issuer): REQUIRED. Same definition as `iss` claim in [](#status-list-token-jwt).
* `2` (subject): REQUIRED. Same definition as `sub` claim in [](#status-list-token-jwt).
* `6` (issued at): REQUIRED. Same definition as `iat` claim in [](#status-list-token-jwt).
* `4` (expiration time): OPTIONAL. Same definition as `exp` claim in [](#status-list-token-jwt).
* `65534` (status list): REQUIRED. The status list claim MUST specify the Status List conforming to the rules outlined in [](#status-list-cbor).

The following additional rules apply:

1. The CWT MAY contain other claims.

2. The CWT MUST be digitally signed using an asymmetric cryptographic algorithm. Relying parties MUST reject the CWT if it is using a Message Authentication Code (MAC) algorithm. Relying parties MUST reject CWTs with an invalid signature.

3. Relying parties MUST reject CWTs that are not valid in all other respects per "CBOR Web Token (CWT)" {{RFC8392}}.

4. Application of additional restrictions and policy are at the discretion of the verifying party.

> The following is a non-normative example for a Status List Token in CWT format:

Reviewer: ```suggestion
The following is a non-normative example for a Status List Token in CWT format (not including the type header yet):
```

}
~~~

## Referenced Token in CWT Format {#referenced-token-cwt}

The Referenced Token MUST be encoded as a "COSE Web Token (CWT)" object according to {{RFC8392}}.

> The following content applies to the JWT Claims Set:

Reviewer: ```suggestion
The following content applies to the CWT Claims Set:
```

* `1` (issuer): REQUIRED. Same definition as `iss` claim in [](#referenced-token-jwt).

> * `65535` (status): REQUIRED. The status claim is encoded as a `Status` CBOR structure and MUST include at least one data item that refers to a status mechanism. Each data item in the `Status` CBOR structure comprises a key-value pair, where the key must be a CBOR text string (Major Type 3) specifying the identifier of the status mechanism, and the corresponding value defines its contents. This specification defines the following data items:

Reviewer: recheck what pre-assigned number to get

* relax requirements on referenced token
* clarify Deflate / zlib compression
* make a reference to the Issuer-Holder-Verifier model of SD-JWT VC

> * add COSE/CWT/BOR encoding

Reviewer: ```suggestion
* add COSE/CWT/CBOR encoding
```
